filters/include/pcl/filters/impl/farthest_point_sampling.hpp

```diff
@@ -0,0 +1,110 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2020-, Open Perception, Inc.
+ *
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder(s) nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#pragma once
+
+#include <pcl/common/geometry.h>
+#include <pcl/filters/farthest_point_sampling.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+#include <algorithm>
+#include <limits>
+#include <random>
+
+template<typename PointT> void
+pcl::FarthestPointSampling<PointT>::applyFilter (std::vector<int> &indices)
+{
+  const std::size_t size = input_->size();
+  //if requested number of point is equal to the point cloud size, copy original cloud
+  if (sample_size_ == size)
+  {
+    indices = *indices_;
+    removed_indices_->clear ();
+    return;
+  }
+  //check if requested number of points is greater than the point cloud size
+  if (sample_size_ > size)
+  {
+    PCL_THROW_EXCEPTION (BadArgumentException,
+                        "Requested number of points is greater than point cloud size!");
+    return;
+  }
+
+  std::vector<float> distances_to_selected_points (size, std::numeric_limits<float>::max ());
+  
+  //set random seed
+  std::mt19937 random_gen(seed_);
+  std::uniform_int_distribution<std::size_t> dis(0, size -1);
+
+  //pick the first point at random
+  int max_index = dis(random_gen);
```

Comment:
`max_index` should be `size_t`

Reply:
`index_t`

Reply:
What's the strategy then: represent both sizes and indexes with index_t? Just want to make sure we agree on something.

Reply:
`index_t` is the type of the index, aka addressable parts of the point cloud
```cpp
if constexpr(std::is_same_v<std::size_t, pcl::index_t>) {
  // Everything works, but might take more memory and be a little slower
}
else {
  // PCL might be unable to address some parts of your point clouds
}
```

Reply:
It does not clearly answer my question but it leads me to believe you implied:
- use index_t for indices
- use size_t for size

Is that it?

Reply:
The indices themselves are a size, because the largest index possible is the size of the point cloud. As such the type used for indices is the smallest safe type that PCL promises and the largest size PCL can safely handle. Using `size_t` isn't bad, but if all you're doing is looking into a point cloud or indices, `index_t` is a better option.

In this particular case, `max_index` is guaranteed to not be more than `std::numeric_limits<pcl::index_t>::max()`, so using `index_t` is better.

Using `std::size_t` for size isn't wrong, it's correct in some situation and over-promising in others. Eg: for `auto num_points = vec.size()`, `auto` can be safely replaced by `std::size_t` (the default safe behavior) or `index_t` (if somewhere we're sure that the size is constrained due to PCL)

Reply:
With this response in mind, the best way to address it is to make everything index_t consistent.